1110-i386-txt-Add-Intel-TXT-core-implementation.patch

```diff
@@ -0,0 +1,1024 @@
++ *     copyright notice, this list of conditions and the following
++ *     disclaimer in the documentation and/or other materials provided
++ *     with the distribution.
++ *   * Neither the name of the Intel Corporation nor the names of its
++ *     contributors may be used to endorse or promote products derived
++ *     from this software without specific prior written permission.
++ *
++ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
++ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
++ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
++ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
++ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
++ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
++ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
++ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
++ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
++ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
++ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
++ * OF THE POSSIBILITY OF SUCH DAMAGE.
++ */
++
++/*
++ *  GRUB  --  GRand Unified Bootloader
++ *  Copyright (C) 2020  Oracle and/or its affiliates.
++ *
++ *  GRUB is free software: you can redistribute it and/or modify
++ *  it under the terms of the GNU General Public License as published by
++ *  the Free Software Foundation, either version 3 of the License, or
++ *  (at your option) any later version.
++ *
++ *  GRUB is distributed in the hope that it will be useful,
++ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
++ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
++ *  GNU General Public License for more details.
++ *
++ *  You should have received a copy of the GNU General Public License
++ *  along with GRUB.  If not, see <http://www.gnu.org/licenses/>.
++ */
++
++#include <grub/loader.h>
++#include <grub/memory.h>
++#include <grub/normal.h>
++#include <grub/err.h>
++#include <grub/misc.h>
++#include <grub/types.h>
++#include <grub/dl.h>
++#include <grub/acpi.h>
++#include <grub/cpu/relocator.h>
++#include <grub/i386/cpuid.h>
++#include <grub/i386/msr.h>
++#include <grub/i386/crfr.h>
++#include <grub/i386/txt.h>
++#include <grub/i386/linux.h>
++#include <grub/i386/memory.h>
++#include <grub/i386/slaunch.h>
++#include <grub/i386/tpm.h>
++
++#define OS_SINIT_DATA_TPM_12_VER	6
++#define OS_SINIT_DATA_TPM_20_VER	7
++
++#define OS_SINIT_DATA_MIN_VER		OS_SINIT_DATA_TPM_12_VER
++
++static grub_err_t
++enable_smx_mode (void)
++{
++  grub_uint32_t caps;
++  grub_uint64_t feat_ctrl = grub_rdmsr (GRUB_MSR_X86_FEATURE_CONTROL);
++
++  if (!(feat_ctrl & GRUB_MSR_X86_FEATURE_CTRL_LOCK))
++    {
++      grub_dprintf ("slaunch", "Firmware didn't lock FEATURE_CONTROL MSR,"
++		    "locking it now\n");
++      feat_ctrl |= GRUB_MSR_X86_FEATURE_CTRL_LOCK;
++    }
```

Comment:
Right, overlooked it. I've added setting VMX only, as SMX is checked earlier. Lack of SMX at this point would most likely mean that the firmware didn't do required TXT initialization and whole process would fail anyway.

Reply:
Is there anything that requires the firmware to enable these knobs to perform initialization? Doesn't it just need to set `TXT.BASE`?

I don't think we have anything to lose by enabling these bits too. Is there a codepath that needs to do `verify_platform` but not `enable_smx_mode`?